retrofit/src/main/java/retrofit2/Retrofit.java

```diff
@@ -64,7 +63,19 @@
  * @author Jake Wharton (jw@squareup.com)
  */
 public final class Retrofit {
-  private final Map<Method, ServiceMethod<?>> serviceMethodCache = new ConcurrentHashMap<>();
+  /**
+   * Method associations in this map will be in one of three states, and may only progress forward
+   * to higher-numbered states.
+   * <ol>
+   *   <li>No value - no one has started or completed parsing annotations for the method.</li>
+   *   <li>Lock object - a thread has started parsing annotations on the method. Once the lock is
+   *   available the map will have been updated with the parsed model.</li>
+   *   <li>{@code ServiceMethod} - annotations for the method have been fully parsed.</li>
+   * </ol>
+   * This map should only be accessed through {@link #loadServiceMethod} which contains the state
+   * transition logic.
+   */
+  private final ConcurrentHashMap<Method, Object> serviceMethodCache = new ConcurrentHashMap<>();
```

Comment:
Concrete type provides access to `putIfAbsent` from API 1 (as opposed to 24 where it moved to `Map`).